README.md
# BERT(S) for Relation Extraction in the Oil and Gas domain

We forked the repository from plkmo/BERT-Relation-Extraction and we are adapting it for the specific domain of oil and gas corpora in the Portuguese language.

## Overview

A PyTorch implementation of the models for the paper ["Matching the Blanks: Distributional Similarity for Relation Learning"](https://arxiv.org/pdf/1906.03158.pdf) published in ACL 2019.  
Note: This is not an official repo for the paper.  
Additional models for relation extraction, implemented here based on the paper's methodology:  
- ALBERT (https://arxiv.org/abs/1909.11942)   
- BioBERT (https://arxiv.org/abs/1901.08746)

For more conceptual details on the implementation, please see https://towardsdatascience.com/bert-s-for-relation-extraction-in-nlp-2c7c3ab487c4

## Requirements
Requirements: Python (3.6+), PyTorch (1.2.0+), Spacy (2.1.8+)  

Pre-trained BERT models (ALBERT, BERT) courtesy of HuggingFace.co (https://huggingface.co)   
Pre-trained BioBERT model courtesy of https://github.com/dmis-lab/biobert   

To use BioBERT(biobert_v1.1_pubmed), download & unzip the [contents](https://drive.google.com/file/d/1zKTBqqrCGlclb3zgBGGpq_70Fx-qFpiU/view?usp=sharing) to ./additional_models folder.   

## Training by matching the blanks (BERT<sub>EM</sub> + MTB)
Run main_pretraining.py with arguments below. Pre-training data can be any .txt continuous text file.  
We use Spacy NLP to grab pairwise entities (within a window size of 40 tokens length) from the text to form relation statements for pre-training. Entities recognition are based on NER and dependency tree parsing of objects/subjects.  

The pre-training data taken from CNN dataset (cnn.txt) that I've used can be downloaded [here.](https://drive.google.com/file/d/1aMiIZXLpO7JF-z_Zte3uH7OCo4Uk_0do/view?usp=sharing)   
However, do note that the paper uses wiki dumps data for MTB pre-training which is much larger than the CNN dataset.   

Note: Pre-training can take a long time, depending on available GPU. It is possible to directly fine-tune on the relation-extraction task and still get reasonable results, following the section below.  

```bash
main_pretraining.py [-h] 
	[--pretrain_data TRAIN_PATH] 
	[--batch_size BATCH_SIZE]
	[--freeze FREEZE]  
	[--gradient_acc_steps GRADIENT_ACC_STEPS]
	[--max_norm MAX_NORM]
	[--fp16 FP_16]  
	[--num_epochs NUM_EPOCHS]
	[--lr LR]
	[--model_no MODEL_NO (0: BERT ; 1: ALBERT ; 2: BioBERT)]  
	[--model_size MODEL_SIZE (BERT: 'bert-base-uncased', 'bert-large-uncased';   
				ALBERT: 'albert-base-v2', 'albert-large-v2';   
				BioBERT: 'bert-base-uncased' (biobert_v1.1_pubmed))]
```

## Fine-tuning on SemEval2010 Task 8 (BERT<sub>EM</sub>/BERT<sub>EM</sub> + MTB)
Run main_task.py with arguments below. Requires SemEval2010 Task 8 dataset, available [here.](https://github.com/sahitya0000/Relation-Classification/blob/master/corpus/SemEval2010_task8_all_data.zip) Download & unzip to ./data/ folder.

```bash
main_task.py [-h] 
	[--train_data TRAIN_DATA]
	[--test_data TEST_DATA]
	[--use_pretrained_blanks USE_PRETRAINED_BLANKS]
	[--num_classes NUM_CLASSES] 
	[--batch_size BATCH_SIZE]
	[--gradient_acc_steps GRADIENT_ACC_STEPS]
	[--max_norm MAX_NORM]
	[--fp16 FP_16]  
	[--num_epochs NUM_EPOCHS]
	[--lr LR]
	[--model_no MODEL_NO (0: BERT ; 1: ALBERT ; 2: BioBERT)]  
	[--model_size MODEL_SIZE (BERT: 'bert-base-uncased', 'bert-large-uncased';   
				ALBERT: 'albert-base-v2', 'albert-large-v2';   
				BioBERT: 'bert-base-uncased' (biobert_v1.1_pubmed))]    
	[--train TRAIN]
	[--infer INFER]
```

### Inference (--infer=1)
To infer a sentence, you can annotate entity1 & entity2 of interest within the sentence with their respective entities tags [E1], [E2]. 
Example:
```bash
Type input sentence ('quit' or 'exit' to terminate):
The surprise [E1]visit[/E1] caused a [E2]frenzy[/E2] on the already chaotic trading floor.

Sentence:  The surprise [E1]visit[/E1] caused a [E2]frenzy[/E2] on the already chaotic trading floor.
Predicted:  Cause-Effect(e1,e2) 
```

```python
from src.tasks.infer import infer_from_trained

inferer = infer_from_trained(args, detect_entities=False)
test = "The surprise [E1]visit[/E1] caused a [E2]frenzy[/E2] on the already chaotic trading floor."
inferer.infer_sentence(test, detect_entities=False)
```
```bash
Sentence:  The surprise [E1]visit[/E1] caused a [E2]frenzy[/E2] on the already chaotic trading floor.
Predicted:  Cause-Effect(e1,e2) 
```

The script can also automatically detect potential entities in an input sentence, in which case all possible relation combinations are inferred:
```python
inferer = infer_from_trained(args, detect_entities=True)
test2 = "After eating the chicken, he developed a sore throat the next morning."
inferer.infer_sentence(test2, detect_entities=True)
```
```bash
Sentence:  [E2]After eating the chicken[/E2] , [E1]he[/E1] developed a sore throat the next morning .
Predicted:  Other 

Sentence:  After eating the chicken , [E1]he[/E1] developed [E2]a sore throat[/E2] the next morning .
Predicted:  Other 

Sentence:  [E1]After eating the chicken[/E1] , [E2]he[/E2] developed a sore throat the next morning .
Predicted:  Other 

Sentence:  [E1]After eating the chicken[/E1] , he developed [E2]a sore throat[/E2] the next morning .
Predicted:  Other 

Sentence:  After eating the chicken , [E2]he[/E2] developed [E1]a sore throat[/E1] the next morning .
Predicted:  Other 

Sentence:  [E2]After eating the chicken[/E2] , he developed [E1]a sore throat[/E1] the next morning .
Predicted:  Cause-Effect(e2,e1) 
```

## FewRel Task
Download the FewRel 1.0 dataset [here.](https://drive.google.com/drive/folders/1ljobnuzxStFQJSlN4ZHMcMhZtEYaRAHy?usp=sharing) and unzip to ./data/ folder.  
Run main_task.py with argument 'task' set as 'fewrel'.
```bash
python main_task.py --task fewrel
```
Results:  
(5-way 1-shot)  
BERT<sub>EM</sub> without MTB, not trained on any FewRel data  
| Model size | Accuracy (41646 samples) |
|------------|--------------------------|
| bert-base-uncased  | 62.229 %         |
| bert-large-uncased | 72.766 %         |


## Benchmark Results

### SemEval2010 Task 8
1) Base architecture: BERT base uncased (12-layer, 768-hidden, 12-heads, 110M parameters)

Without MTB pre-training: F1 results when trained on 100 % training data:
![](https://github.com/plkmo/BERT-Relation-Extraction/blob/master/results/CNN/task_test_f1_vs_epoch_0.png) 


2) Base architecture: ALBERT base uncased (12 repeating layers, 128 embedding, 768-hidden, 12-heads, 11M parameters)  

Without MTB pre-training: F1 results when trained on 100 % training data:
![](https://github.com/plkmo/BERT-Relation-Extraction/blob/master/results/CNN/task_test_f1_vs_epoch_1.png) 

## To add
- inference & results on benchmarks (SemEval2010 Task 8) with MTB pre-training
- felrel task



history.txt
    1  2021-08-23 09:40:59 nvidia-smi
    2  2021-08-23 09:43:41 df -h
    3  2021-08-23 09:44:24 conda
    4  2021-08-23 09:44:40 singularity 
    5  2021-08-23 09:46:17 wget www.uol.com.br
    6  2021-08-23 09:46:25 curl www.uol.com.br
    7  2021-08-23 09:47:41 env
    8  2021-08-23 09:47:53 exit
    9  2021-08-23 09:52:12 nvidia-smi
   10  2021-08-23 10:21:16 scp u4vn@npaa8749.petrobras.biz:~/.bashrc .
   11  2021-08-23 10:22:07 exit
   12  2021-08-23 10:22:27 df -h
   13  2021-08-23 10:22:35 mkdir /home/u4vn
   14  2021-08-23 10:22:39 cd /home
   15  2021-08-23 10:22:40 ls
   16  2021-08-23 10:22:43 cd U4VN
   17  2021-08-23 10:22:45 LS
   18  2021-08-23 10:22:47 ls
   19  2021-08-23 10:23:25 wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
   20  2021-08-23 10:23:36 curl https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
   21  2021-08-23 10:25:04 yum install wget
   22  2021-08-23 10:26:56 mkdir Downloads
   23  2021-08-23 10:26:58 cd Downloads/
   24  2021-08-23 10:27:34 scp u4vn@mi00311323x.petrobras.biz:Miniconda3-latest-Linux-x86_64.sh . 
   25  2021-08-23 10:27:44 ls
   26  2021-08-23 10:27:54 chmod +x Miniconda3-latest-Linux-x86_64.sh 
   27  2021-08-23 10:27:57 ./Miniconda3-latest-Linux-x86_64.sh 
   28  2021-08-23 10:28:49 source ~/.bashrc
   29  2021-08-23 10:30:47 conda install wget
   30  2021-08-23 10:30:58 wget
   31  2021-08-23 10:33:48 exit
   32  2021-08-24 21:14:33 conda activate nlp
   33  2021-08-24 21:14:41 python
   34  2021-08-24 21:15:27 ls
   35  2021-08-24 21:15:30 cd BERT-Relation-Extraction/
   36  2021-08-24 21:15:30 ls
   37  2021-08-25 08:04:21 python
   38  2021-08-25 08:04:28 conda activate nlp
   39  2021-08-25 08:04:30 conda env list
   40  2021-08-25 08:59:52 ls
   41  2021-08-25 08:59:56 ps aux
   42  2021-08-25 09:00:03 nvidia-smi
   43  2021-08-25 09:01:04 wget
   44  2021-08-25 09:01:09 conda install wget
   45  2021-08-25 09:01:58 cd BERT-Relation-Extraction/
   46  2021-08-25 09:01:59 cd data
   47  2021-08-25 09:02:08 wget https://drive.google.com/file/d/1aMiIZXLpO7JF-z_Zte3uH7OCo4Uk_0do/view?usp=sharing
   48  2021-08-25 09:03:26 cd ..
   49  2021-08-25 09:04:35 python main_pretraining.py --pretrain_data ./data/cnn.txt 
   50  2021-08-25 09:04:46 conda install matplotlib
   51  2021-08-25 10:16:03 conda activate nlp
   52  2021-08-25 10:16:18 python -m spacy download pt_core_news_lg
   53  2021-08-25 10:17:51 exit
   54  2021-08-25 14:28:30 ls -la
   55  2021-08-25 14:54:56 cd ..
   56  2021-08-25 14:54:58 cd transformers_cache/
   57  2021-08-25 14:54:59 ls
   58  2021-08-25 08:04:21 python
   59  2021-08-25 08:04:28 conda activate nlp
   60  2021-08-25 08:04:30 conda env list
   61  2021-08-25 08:59:52 ls
   62  2021-08-25 08:59:56 ps aux
   63  2021-08-25 09:00:03 nvidia-smi
   64  2021-08-25 09:01:04 wget
   65  2021-08-25 09:01:09 conda install wget
   66  2021-08-25 09:01:58 cd BERT-Relation-Extraction/
   67  2021-08-25 09:01:59 cd data
   68  2021-08-25 09:02:08 wget https://drive.google.com/file/d/1aMiIZXLpO7JF-z_Zte3uH7OCo4Uk_0do/view?usp=sharing
   69  2021-08-25 09:03:26 cd ..
   70  2021-08-25 09:04:35 python main_pretraining.py --pretrain_data ./data/cnn.txt 
   71  2021-08-25 09:04:46 conda install matplotlib
   72  2021-08-25 09:06:29 python main_pretraining.py --pretrain_data ./data/cnn.txt 
   73  2021-08-25 09:07:05 python -m spacy download en_core_web_lg
   74  2021-08-25 09:09:02 python main_pretraining.py --pretrain_data ./data/cnn.txt 
   75  2021-08-25 09:12:24 conda install -c conda-forge boto3
   76  2021-08-25 09:13:54 python main_pretraining.py --pretrain_data ./data/cnn.txt 
   77  2021-08-25 09:28:05 vim ~/.bashrc
   78  2021-08-25 09:28:14 cd ..
   79  2021-08-25 09:28:18 ls
   80  2021-08-25 09:28:29 mkdir transformers_cache
   81  2021-08-25 09:28:32 cd transformers_cache/
   82  2021-08-25 09:28:34 pwd
   83  2021-08-25 09:28:41 vim ~/.bashrc
   84  2021-08-25 09:31:51 wget https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-uncased-pytorch_model.bin
   85  2021-08-25 09:45:07 ls
   86  2021-08-25 09:45:15 vim ~/.bashrc
   87  2021-08-25 09:45:33 echo $TRANSFORMERS_CACHE
   88  2021-08-25 09:45:38 pwd
   89  2021-08-25 09:45:54 cd ..
   90  2021-08-25 09:45:58 cd git
   91  2021-08-25 09:46:00 cd BERT-Relation-Extraction/
   92  2021-08-25 09:46:13 python main_pretraining.py --pretrain_data ./data/cnn.txt 
   93  2021-08-25 09:46:21 conda activate nlp
   94  2021-08-25 09:46:23 python main_pretraining.py --pretrain_data ./data/cnn.txt 
   95  2021-08-25 13:54:25 python main_pretraining.py
   96  2021-08-25 14:07:09 python main_pretraining.py --model-size 'neuralmind/bert-base-portuguese-cased'
   97  2021-08-25 14:07:30 python main_pretraining.py --model_size 'neuralmind/bert-base-portuguese-cased'
   98  2021-08-25 14:58:29 python main_pretraining.py --model-size 'neuralmind/bert-base-portuguese-cased'
   99  2021-08-25 14:58:38 python main_pretraining.py --model_size 'bert-base-multilingual-cased'
  100  2021-08-25 15:19:50 python main_pretraining.py --model-size 'neuralmind/bert-base-portuguese-cased'
  101  2021-08-25 15:20:02 python main_pretraining.py --model_size 'neuralmind/bert-base-portuguese-cased'
  102  2021-08-25 14:09:41 conda activate nlp
  103  2021-08-25 14:09:47 python
  104  2021-08-25 14:15:36 cd BERT-Relation-Extraction/
  105  2021-08-25 14:15:37 ls
  106  2021-08-25 14:15:39 ls -la
  107  2021-08-25 14:15:42 cd ..
  108  2021-08-25 14:15:43 ls
  109  2021-08-25 14:15:44 ls -la
  110  2021-08-25 14:16:05 mv bert-base-uncased-pytorch_model.bin ~/transformers_cache/
  111  2021-08-25 14:16:16 cd ..
  112  2021-08-25 14:16:17 ls
  113  2021-08-25 14:16:20 cd transformers_cache/
  114  2021-08-25 14:16:21 ls
  115  2021-08-25 14:17:12 cd /home/U4VN/.cache/torch/transformers/
  116  2021-08-25 14:17:14 ls
  117  2021-08-25 14:17:19 ls -la

main_task.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon Dec  2 17:40:16 2019

@author: weetee
"""
from src.tasks.trainer import train_and_fit
from src.tasks.infer import infer_from_trained, FewRel
import logging
from argparse import ArgumentParser

'''
This fine-tunes the BERT model on SemEval, FewRel tasks
'''

logging.basicConfig(format='%(asctime)s [%(levelname)s]: %(message)s', \
                    datefmt='%m/%d/%Y %I:%M:%S %p', level=logging.INFO)
logger = logging.getLogger('__file__')

if __name__ == "__main__":
    parser = ArgumentParser()
    parser.add_argument("--task", type=str, default='semeval', help='semeval, fewrel')
    parser.add_argument("--train_data", type=str, default='./data/SemEval2010_task8_all_data/SemEval2010_task8_training/TRAIN_FILE.TXT', \
                        help="training data .txt file path")
    parser.add_argument("--test_data", type=str, default='./data/SemEval2010_task8_all_data/SemEval2010_task8_testing_keys/TEST_FILE_FULL.TXT', \
                        help="test data .txt file path")
    parser.add_argument("--use_pretrained_blanks", type=int, default=0, help="0: Don't use pre-trained blanks model, 1: use pre-trained blanks model")
    parser.add_argument("--num_classes", type=int, default=19, help='number of relation classes')
    parser.add_argument("--batch_size", type=int, default=32, help="Training batch size")
    parser.add_argument("--gradient_acc_steps", type=int, default=2, help="No. of steps of gradient accumulation")
    parser.add_argument("--max_norm", type=float, default=1.0, help="Clipped gradient norm")
    parser.add_argument("--fp16", type=int, default=0, help="1: use mixed precision ; 0: use floating point 32") # mixed precision doesn't seem to train well
    parser.add_argument("--num_epochs", type=int, default=11, help="No of epochs")
    parser.add_argument("--lr", type=float, default=0.00007, help="learning rate")
    parser.add_argument("--model_no", type=int, default=0, help='''Model ID: 0 - BERT\n
                                                                            1 - ALBERT\n
                                                                            2 - BioBERT''')
    parser.add_argument("--model_size", type=str, default='bert-base-uncased', help="For BERT: 'bert-base-uncased', \
                                                                                                'bert-large-uncased',\
                                                                                    For ALBERT: 'albert-base-v2',\
                                                                                                'albert-large-v2'\
                                                                                    For BioBERT: 'bert-base-uncased' (biobert_v1.1_pubmed)")
    parser.add_argument("--train", type=int, default=1, help="0: Don't train, 1: train")
    parser.add_argument("--infer", type=int, default=1, help="0: Don't infer, 1: Infer")
    
    args = parser.parse_args()
    
    if (args.train == 1) and (args.task != 'fewrel'):
        net = train_and_fit(args)
        
    if (args.infer == 1) and (args.task != 'fewrel'):
        inferer = infer_from_trained(args, detect_entities=True)
        
        test2 = "O [E1]campo de Marlim[/E1] é o maior campo da [E2]Bacia de Campos[/E2]."
        inferer.infer_sentence(test2, detect_entities=False)
        
        test2 = "Os campos de [E1]Marlim[/E1] e [E2]Albacora[/E2] são os maiores produtores da região."
        inferer.infer_sentence(test2, detect_entities=False)
        
        test2 = "O estado mais populoso do [E1]Brasil[/E2] é o estado de [E2]São Paulo[/E2], com uma população maior do que muitos países europeus."
        inferer.infer_sentence(test2, detect_entities=False)
        
        while True:
            sent = input("Type input sentence ('quit' or 'exit' to terminate):\n")
            if sent.lower() in ['quit', 'exit']:
                break
            inferer.infer_sentence(sent, detect_entities=False)
    
    if args.task == 'fewrel':
        fewrel = FewRel(args)
        meta_input, e1_e2_start, meta_labels, outputs = fewrel.evaluate()